crates/fs/private/hamt/node.rs

```diff
@@ -0,0 +1,251 @@
+// TODO(appcypher): Based on ipld_hamt implementation
+
+use std::rc::Rc;
+
+use crate::{BlockStore, Link};
+use anyhow::Result;
+use async_recursion::async_recursion;
+use bitvec::array::BitArray;
+use bitvec::bitarr;
+use bitvec::order::Lsb0;
+use serde::{
+    de::{Deserialize, DeserializeOwned},
+    Deserializer, Serialize, Serializer,
+};
+use sha3::Sha3_256;
+
+use super::{
+    hash::{GenerateHash, HashQuartets},
+    Pair, Pointer, HAMT_BITMASK_BYTES, HAMT_BITMASK_SIZE,
+};
+
+//--------------------------------------------------------------------------------------------------
+// Type Definitions
+//--------------------------------------------------------------------------------------------------
+
+pub type BitMaskType = [u8; HAMT_BITMASK_BYTES];
+
+#[derive(Debug, Clone, PartialEq)]
+pub struct Node<K, V> {
+    pub(crate) bitmask: BitArray<BitMaskType>,
+    pub(crate) pointers: Vec<Pointer<K, V>>,
+}
+
+//--------------------------------------------------------------------------------------------------
+// Implementations
+//--------------------------------------------------------------------------------------------------
+
+impl<K, V> Node<K, V>
+where
+    K: DeserializeOwned + Serialize + AsRef<[u8]> + Clone + Eq,
+    V: DeserializeOwned + Serialize + Clone,
+{
+    pub async fn set<B: BlockStore>(
+        self: Rc<Self>,
+        key: K,
+        value: V,
+        store: &mut B,
+    ) -> Result<Rc<Self>> {
+        let hash = &Sha3_256::generate_hash(&key);
+        self.modify_value(&mut HashQuartets::new(hash), key, value, store)
+            .await
+    }
+
+    pub async fn get<'a, B: BlockStore>(
+        self: &'a Rc<Self>,
+        key: &K,
+        store: &B,
+    ) -> Result<Option<&'a V>> {
+        Ok(self.search(key, store).await?.map(|pair| &pair.value))
+    }
+
+    /// Recursively resolves all nodes under the directory to Cids.
+    #[async_recursion(?Send)]
+    pub async fn flush<'a, B: BlockStore>(self: &'a Rc<Self>, store: &mut B) -> Result<()> {
```

Comment:
We should be able to implement `AsyncSerialize` for the HAMT nodes now instead :+1: